In [1]:
# uncomment the following line to install the pandas library
!pip install pandas 

'''Pandas is used to gather data sets through its DataFrames implementation'''
import pandas as pd

# Popcorn Hack 1

In [2]:
import pandas as pd

# Example student data
student_data = pd.DataFrame({
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Emma', 'Fiona', 'George'],
    'Score': [95, 88, 76, 92, 84, 67, 89],
    'Grade': ['A', 'B', 'C', 'A', 'B', 'D', 'B']
})

# print(student_data)

def find_students_in_range(df, min_score, max_score):
    return df[(df['Score'] >= min_score) & (df['Score'] <= max_score)]

# Test
result = find_students_in_range(student_data, 80, 90)
print(result)

     Name  Score Grade
1     Bob     88     B
4    Emma     84     B
6  George     89     B


# Popcorn Hack 2

In [3]:
def add_letter_grades(df):
    def letter(score):
        if score >= 90:
            return 'A'
        elif score >= 80:
            return 'B'
        elif score >= 70:
            return 'C'
        elif score >= 60:
            return 'D'
        else:
            return 'F'
    
    df['Letter'] = df['Score'].apply(letter)
    return df

# Test
graded = add_letter_grades(student_data)
print(graded)


      Name  Score Grade Letter
0    Alice     95     A      A
1      Bob     88     B      B
2  Charlie     76     C      C
3    David     92     A      A
4     Emma     84     B      B
5    Fiona     67     D      D
6   George     89     B      B


# Popcorn Hack 3

In [4]:
def find_mode(series):
    return series.mode().iloc[0] if not series.mode().empty else None

# Test
mode = find_mode(pd.Series([1, 2, 2, 3, 4, 2, 5]))
print("Mode:", mode)


Mode: 2


---

## General Setup

In [13]:
import pandas as pd

smartpark_df = pd.read_csv('/home/mihirunix/nighthawk/MihirCSP/datasets/treas_parking_meters_loc_datasd.csv')
smartpark_df.head()

zone             area        sub_area    pole  config_id  \
0  Downtown             Core  1000 FIRST AVE  1-1004      49382   
1  Downtown  Core - Columbia  1000 FIRST AVE  1-1004       9000   
2  Downtown             Core  1000 FIRST AVE  1-1006      49382   
3  Downtown  Core - Columbia  1000 FIRST AVE  1-1006       9000   
4  Downtown             Core  1000 FIRST AVE  1-1008      49382   

                                 config_name       date_inventory        lat  \
0                                Sunday Mode  2021-01-04 00:00:00  32.715904   
1  2 Hour Max $1.25 HR 8am-6pm Mon-Sat (NFC)  2018-11-11 00:00:00  32.715904   
2                                Sunday Mode  2021-01-04 00:00:00  32.716037   
3  2 Hour Max $1.25 HR 8am-6pm Mon-Sat (NFC)  2018-11-11 00:00:00  32.716037   
4                                Sunday Mode  2021-01-04 00:00:00  32.716169   

          lng      sapid  
0 -117.163929  SS-000031  
1 -117.163929  SS-000031  
2 -117.163930  SS-000031  
3 -117.163930  SS-000031  
4 -117.163931  SS-000031

## ✅ Hack 1: Find Configurations Older than a Certain Year

Find all entries in the dataset prior to 2019.

In [14]:
smartpark_df['date_inventory'] = pd.to_datetime(smartpark_df['date_inventory'])
older_configs = smartpark_df[smartpark_df['date_inventory'].dt.year < 2019]
older_configs.head()

zone             area        sub_area    pole  config_id  \
1  Downtown  Core - Columbia  1000 FIRST AVE  1-1004       9000   
3  Downtown  Core - Columbia  1000 FIRST AVE  1-1006       9000   
5  Downtown  Core - Columbia  1000 FIRST AVE  1-1008       9000   
7  Downtown  Core - Columbia  1000 FIRST AVE  1-1020       9115   
9  Downtown  Core - Columbia  1300 FIRST AVE  1-1310      16516   

                                         config_name date_inventory  \
1          2 Hour Max $1.25 HR 8am-6pm Mon-Sat (NFC)     2018-11-11   
3          2 Hour Max $1.25 HR 8am-6pm Mon-Sat (NFC)     2018-11-11   
5          2 Hour Max $1.25 HR 8am-6pm Mon-Sat (NFC)     2018-11-11   
7          15 Min Max $1.25 HR 8am-6pm Mon-Sat (NFC)     2018-11-11   
9  2 Hour Max $1.25 HR 8am-4pm Mon-Fri 4pm-6pm TO...     2018-11-11   

         lat         lng      sapid  
1  32.715904 -117.163929  SS-000031  
3  32.716037 -117.163930  SS-000031  
5  32.716169 -117.163931  SS-000031  
7  32.717890 -117.161278  SS-000031  
9  32.719024 -117.163951  SS-000028


---

## ✅ Hack 2: Count Configs by Area

In [15]:
configs_by_area = smartpark_df.groupby('area')['config_id'].count().reset_index(name='Config Count')
configs_by_area.head()

area  Config Count
0  Bankers Hill          1787
1  Barrio Logan            69
2   CBD - Trial            19
3       College            10
4      Columbia           163


---

## ✅ Hack 3: Find the Most Common Parking Configuration


In [16]:
most_common_config = smartpark_df['config_name'].mode().iloc[0]
print("Most common config:", most_common_config)

Most common config: Sunday Mode



---

## ✅ Hack 4: Compare Configs per Zone Over Time

In [17]:
smartpark_df['year'] = smartpark_df['date_inventory'].dt.year
configs_by_zone_year = smartpark_df.groupby(['zone', 'year']).size().reset_index(name='Config Count')
configs_by_zone_year.head()

zone  year  Config Count
0      City  2018            18
1      City  2019            62
2      City  2021            75
3  Downtown  2018          1814
4  Downtown  2019          1486


---

## ✅ Hack 5: Count Unique GPS Locations

In [18]:
unique_locations = smartpark_df[['lat', 'lng']].drop_duplicates().shape[0]
print("Unique parking locations (lat/lng pairs):", unique_locations)

Unique parking locations (lat/lng pairs): 5308



---

## 🧠 Summary

- Oldest config is before 2019: `"2 Hour Max $1.25 HR 8am-6pm Mon-Sat (NFC)"`
- Most common configuration: `"Sunday Mode"`
- Total unique GPS locations: `5308`
- Grouping, filtering, and traversal were performed using Pandas!